src/Workspaces/Core/Portable/EmbeddedLanguages/Common/EmbeddedSyntaxToken.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Common
+{
+    internal struct EmbeddedSyntaxToken<TSyntaxKind> where TSyntaxKind : struct
+    {
+        public readonly TSyntaxKind Kind;
+        public readonly ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> LeadingTrivia;
+        public readonly ImmutableArray<VirtualChar> VirtualChars;
+        public readonly ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> TrailingTrivia;
+        internal readonly ImmutableArray<EmbeddedDiagnostic> Diagnostics;
+        public readonly object Value;
+
+        public EmbeddedSyntaxToken(
+            TSyntaxKind kind,
+            ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> leadingTrivia,
+            ImmutableArray<VirtualChar> virtualChars,
+            ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> trailingTrivia,
+            ImmutableArray<EmbeddedDiagnostic> diagnostics, object value)
+        {
+            Kind = kind;
+            LeadingTrivia = leadingTrivia;
+            VirtualChars = virtualChars;
+            TrailingTrivia = trailingTrivia;
+            Diagnostics = diagnostics;
+            Value = value;
+        }
+
+        public bool IsMissing => VirtualChars.IsEmpty;
+
+        public EmbeddedSyntaxToken<TSyntaxKind> AddDiagnosticIfNone(EmbeddedDiagnostic diagnostic)
```

Comment:
It looks like this method is not used or tested. If this method is removed, then `WithDiagnostics` can probably be removed too.

Reply:
The tests are in another PR because they destroy github.  You can see the tests for regex and json in these PRs:

https://github.com/dotnet/roslyn/pull/25520

I would prefer to keep them separate.

Reply:
Also note: this is hte common embedded languages subsystem part of the work.  There are no explicit tests for this part as this is just an implementation detail of hte final regex parser/features.